Be nicer about possible misconfigurations

thejcannon opened the pull request:
Fixes https://github.com/pypa/setuptools_scm/issues/1003 by turning situations where the user might get an `AssertionError` and turn them into descriptive `ValueError`s.

Additionally, tests have been added to test the behavior.

The situations fixed are (with the real-world facepalm situations I ended up in):
- The entrypoint for `version_scheme` and `local_scheme` don't exist
  - :facepalm: I accidentally used the function name once, instead of the entrypoint name
- The entrypoint(s) for `version_scheme` and `local_scheme` don't return anything
  - :facepalm: I forgot to add a `return` to my custom scheme
- The `tag_regex` is bad, and was loaded from the config
  - :facepalm: This one should be self-explanatory
- The `tag_regex` matched, but the version was blank
  - :facepalm: My regex was borked

con-f-use committed:
cli: extend ls comand for more info

RonnyPfannschmidt committed:
chore: hg file finder - use check=True api

RonnyPfannschmidt committed:
chore: has_command - use default timeout automatically

RonnyPfannschmidt committed:
fix #957 - add subprocess timeout

RonnyPfannschmidt committed:
chore: add exception info to warning logging

RonnyPfannschmidt committed:
begin to address #993 - add normcase

normcase around realpath all over the place should help to operate with normalized paths in general

RonnyPfannschmidt committed:
fixup #957 - timeout passover for git

RonnyPfannschmidt committed:
chore: add py3.12 in tox.ini and add wheel

RonnyPfannschmidt committed:
fix #966 - list all variables for version file templates

RonnyPfannschmidt committed:
fix #960: add a ``--force-write-version-files`` flag for the cli

RonnyPfannschmidt committed:
fix #953 - info log level for missing tool section

this ensures we dont spam projects

RonnyPfannschmidt committed:
fix #993 and add a test

thejcannon committed:
a few fixes and tests

thejcannon committed:
newline

thejcannon committed:
space

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

thejcannon committed:
pre-commit

thejcannon committed:
fix bad string

thejcannon committed:
Add a "only-version" version scheme (#1005)

* Add "just-tag"

* docs too

* rename

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Rename

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Dont autoformat this

* docs warning

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

RonnyPfannschmidt commented on src/setuptools_scm/_config.py:
```diff
@@ -41,9 +41,9 @@ def _check_tag_regex(value: str | Pattern[str] | None) -> Pattern[str]:
-        warnings.warn(
-            "Expected tag_regex to contain a single match group or a group named"
-            " 'version' to identify the version part of any tag."
+        raise ValueError(
```
this one is possibly a breaking change im not sure we can integrate it off hand

RonnyPfannschmidt commented on src/setuptools_scm/_config.py:
```diff
@@ -105,6 +105,9 @@ class Configuration:
     parent: _t.PathT | None = None
 
+    def __post_init__(self) -> None:
+        self.tag_regex = _check_tag_regex(self.tag_regex)
```
taking this here without leaving it in place creates potental trouble for valies valid in config but not in the config objects

RonnyPfannschmidt commented on src/setuptools_scm/_config.py:
```diff
@@ -139,13 +142,11 @@ def from_data(
         given configuration data
         create a config instance after validating tag regex/version class
         """
-        tag_regex = _check_tag_regex(data.pop("tag_regex", None))
```
this check needs to say even if the validation gets more struct

RonnyPfannschmidt reviewed (commented):
i like the idea behind how the details change, but i aslo beleive a larger change is necessary to do the path/version scheme handling in a better way

due to personal time constraints it may take a be a while to propperly reiterate this

thank you for getting the ball rolling and preparing this as a initial vesion

RonnyPfannschmidt committed:
chore: pass correct parameter name fo cli fallback

RonnyPfannschmidt committed:
chore: consistently use kwonly args inside config.from_data

RonnyPfannschmidt committed:
chore: turn empty tag regex warning into constant

RonnyPfannschmidt committed:
test fix: correct cli test after pytest 8 exposed a bug

RonnyPfannschmidt committed:
ci fix: unshallow

RonnyPfannschmidt committed:
chore: drop black, flake8 and reoder-python-imports

RonnyPfannschmidt committed:
fix ruff config

RonnyPfannschmidt committed:
Merge pull request #996 from RonnyPfannschmidt/ronny/cleanups

cleanups + bugfixes

dcermak committed:
Only depend on typing_extensions for Python < 3.10

All required types are part of the stdlib as of Python 3.10

Commit:
fix non-escaped markdown in docs

RonnyPfannschmidt committed:
Merge pull request #1017 from JCGoran/fix_markdown

Fix non-escaped markdown in docs

RonnyPfannschmidt committed:
Merge pull request #1015 from dcermak/typing_extensions-only-for-older-py

Only depend on typing_extensions for Python < 3.10

RonnyPfannschmidt committed:
fix #1018 - allow NonNormalizedVersion in semver

RonnyPfannschmidt committed:
upgrade pre-commit

RonnyPfannschmidt committed:
Merge pull request #1020 from RonnyPfannschmidt/fix-1018-allow-non-normalized-semver

fix #1018 allow non normalized semver

RonnyPfannschmidt committed:
Merge pull request #1021 from RonnyPfannschmidt/pre-commit-upgrade

pre commit upgrade

RonnyPfannschmidt committed:
Apply suggestions from code review

RonnyPfannschmidt committed:
Merge branch 'main' into main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

RonnyPfannschmidt committed:
fix ruff warning on concat

RonnyPfannschmidt committed:
Merge pull request #959 from con-f-use/main

add 'info' command to cli

RonnyPfannschmidt commented:
i intend to merge this after hte next minor release as its a "breaking change" that should go in nonetheless

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.3.0 → v0.3.2](https://github.com/astral-sh/ruff-pre-commit/compare/v0.3.0...v0.3.2)
- [github.com/pre-commit/mirrors-mypy: v1.8.0 → v1.9.0](https://github.com/pre-commit/mirrors-mypy/compare/v1.8.0...v1.9.0)

RonnyPfannschmidt committed:
Merge pull request #1023 from pypa/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

danchr committed:
Read branch data from Mercurial archives

Mercurial has included branches in `.hg_archival.txt` since about 2009.
To easily test this, I adjusted the archival test to use the
`release-branch-semver` scheme.

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.3.2 → v0.3.3](https://github.com/astral-sh/ruff-pre-commit/compare/v0.3.2...v0.3.3)

RonnyPfannschmidt committed:
Merge pull request #1026 from pypa/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.3.3 → v0.3.4](https://github.com/astral-sh/ruff-pre-commit/compare/v0.3.3...v0.3.4)

RonnyPfannschmidt committed:
Merge pull request #1027 from pypa/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

LecrisUT committed:
Make .git_archival.txt more reproducible

RonnyPfannschmidt committed:
Merge pull request #1033 from LecrisUT/fix/git_archival

Make .git_archival.txt more reproducible

RonnyPfannschmidt committed:
Merge pull request #1024 from danchr/hg-archive-branch

Read branch data from Mercurial archives

ap-- committed:
tests: add git get_head_date test for isoformat dates ending with Z

ap-- committed:
setuptools_scm.git: fix handling of dates in isoformat ending with Z

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

RonnyPfannschmidt committed:
Merge pull request #1039 from ap--/fix-timezone-error

Fix timezone error

RonnyPfannschmidt committed:
bugfix: use extras in tox to ensure installation of test deps

RonnyPfannschmidt committed:
Merge pull request #1043 from RonnyPfannschmidt/fix-tox

bugfix: use extras in tox to ensure installation of test deps

DimitriPapadopoulos committed:
Apply ruff/flake8-implicit-str-concat rule ISC001

- Revert to multiline strings, as before #255 / ac2fee8.
- Guard the multiline strings from black (and now ruff) unfolding
  them back to a single line.

https://docs.astral.sh/ruff/rules/single-line-implicit-string-concatenation/

hugovk committed:
Exclude bots from generated release notes

RonnyPfannschmidt committed:
Merge pull request #1046 from hugovk/exclude-bots

Exclude bots from generated release notes

DimitriPapadopoulos committed:
Apply ruff/pyupgrade rule UP031

	UP031 Use format specifiers instead of percent format

DimitriPapadopoulos committed:
Apply ruff/refurb rule FURB118

	FURB118 Use `operator.itemgetter(slice(7))` instead of defining a function

RonnyPfannschmidt committed:
Merge pull request #1047 from DimitriPapadopoulos/UP_FURB

Apply ruff/pyupgrade (UP) and ruff/refurb (FURB) rules

RonnyPfannschmidt committed:
Merge pull request #1044 from DimitriPapadopoulos/ISC

Apply ruff/flake8-implicit-str-concat rule ISC001

njzjz committed:
support `pipx run setuptools_scm`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

DimitriPapadopoulos committed:
Apply ruff/tryceratops rule TRY401

TRY401 Redundant exception object included in `logging.exception` call

shenxianpeng committed:
Update README.md to correct CI badge status

It seems that the current GitHub CI badge does not correctly reflect the true status.

RonnyPfannschmidt committed:
Merge pull request #1055 from shenxianpeng/patch-1

brettcannon committed:
Fix a link in usage.md

RonnyPfannschmidt committed:
Merge pull request #1057 from brettcannon/patch-1

RonnyPfannschmidt committed:
Merge pull request #1053 from DimitriPapadopoulos/TRY401

jamesbraza committed:
Documented git archive support in README and docs

RonnyPfannschmidt committed:
Merge pull request #1061 from jamesbraza/git-archive-docs

RonnyPfannschmidt committed:
Merge branch 'main' into patch-1

RonnyPfannschmidt committed:
Merge pull request #1051 from njzjz/patch-1

DimitriPapadopoulos committed:
Apply repo-review rules PP306, PP307, PP308

PP306: Specifies strict config
`--strict-config` should be in `addopts = [...]`. This forces
an error if a config setting is misspelled.

PP307: Specifies strict markers
`--strict-markers` should be in `addopts = [...]`. This forces
all markers to be specified in config, avoiding misspellings.

PP308: Specifies useful pytest summary
An explicit summary flag like `-ra` should be in `addopts = [...]`
(print summary of all fails/errors).

DimitriPapadopoulos committed:
Apply repo-review rule PC191

PC191: Ruff show fixes if fixes enabled
If `--fix` is present, `--show-fixes` must be too.

DimitriPapadopoulos committed:
Apply repo-review rule RF003

RF003: src directory specified if used
Must specify `src` directory if it exists.

DimitriPapadopoulos committed:
Apply repo-review rule RF103

RF103: pyupgrade must be selected
Must select the pyupgrade `UP` checks.

DimitriPapadopoulos committed:
Apply ruff rules

DimitriPapadopoulos committed:
Add repo-review pre-commit hook

RonnyPfannschmidt committed:
Merge pull request #1063 from DimitriPapadopoulos/repo-review

Apply assorted repo-review rules

jaraco committed:
Rename the project to setuptools-scm in docs and GitHub.

Closes #1067

jaraco committed:
Updated more references to the legacy name in Python code.

jaraco committed:
Also update URL reference in pyproject.

ianw committed:
docs: clarify usage of version files

I found the current discussion around version files somewhat
confusing.

When you're reading the usage guide, the second point is telling you
to make a version file; strongly implying you should.  But in reality,
you most likely do not need to bother with keeping a specific version
file.  Explain more clearly about the implications of this.

Under this is a section on using `python -m setuptools_scm --help`
which seems like an orphan from the "as a cli" section; move it down
there.

The runtime section starts by telling you what is strongly discouraged
without explaination why.  Make it clearer by giving the first example
as the most supported thing, which is using standard importlib; then
briefly discuss the version file from above, explaining what's in it
for a standard Python template, then the strongly discouraged direct
import method.

In the config doc, update to discuss `importlib` for consistency.

Correct some capitalization and other minor formatting bits while
we're here.

RonnyPfannschmidt committed:
Merge pull request #1068 from jaraco/debt/rename

Rename the project to setuptools-scm

jamesbraza committed:
Updated version.py template to include __all__

jamesbraza committed:
Added a CHANGELOG entry

jamesbraza committed:
Added 'ruff check' unit test

RonnyPfannschmidt committed:
Merge pull request #1076 from jamesbraza/all-in-version

`version.py` including `__all__`

paugier committed:
Better cleanup PYTHONPATH for Mercurial calls (support for PDM)

RonnyPfannschmidt committed:
Merge pull request #1080 from paugier/hg-git_with_pdm

Better cleanup PYTHONPATH for Mercurial calls (support for PDM)

paugier committed:
pipx.run entry-points should be setuptools-scm

mayeut committed:
Add a lower bound pin on setuptools dependency

anchal39 committed:
Fixes #1090. Pinned requirement to tomli<=2.0.2.

Signed-off-by: Anchal Agarwal <anchal@voereir.com>

maresb committed:
Add cross-reference from version_scheme config to implementations

I had a difficult time finding the implementations in the docs, despite knowing of their existence. Adding this cross-reference should help.

x

maresb committed:
Remove the word "local" per PEP 440

In PEP 440, "local" refers to the optional portion of a version after the `+` <https://peps.python.org/pep-0440/#local-version-identifiers>. But `version_scheme` doesn't apply to this part.

maresb committed:
Clarify the definition of "local"

Most people reading this sentence probably don't have PEP 440 memorized.

maresb committed:
Add cross-reference from local_scheme config to implementations

RonnyPfannschmidt committed:
Merge pull request #1094 from maresb/version-scheme-docs

Improve the docs about `version_scheme` and `local_scheme`

RonnyPfannschmidt committed:
Merge pull request #1091 from anchal39/tomli-version-fix

Fixes #1090. Pinned requirement to tomli<=2.0.2.

RonnyPfannschmidt committed:
Merge pull request #1088 from mayeut/patch-1

Add a lower bound pin on setuptools dependency

RonnyPfannschmidt committed:
Merge branch 'main' into version-file

RonnyPfannschmidt committed:
Merge pull request #1069 from ianw/version-file

docs: clarify usage of version files

RonnyPfannschmidt committed:
Merge pull request #1081 from paugier/fix-pipx-command-name

pipx.run entry-points should be setuptools-scm

paugier committed:
Mention SETUPTOOLS_SCM_PRETEND_VERSION_FOR in error message

RonnyPfannschmidt committed:
Merge pull request #1102 from paugier/error-message-env-var

Mention SETUPTOOLS_SCM_PRETEND_VERSION_FOR in error message

RonnyPfannschmidt committed:
merge main

RonnyPfannschmidt commented:
i pushed a merge to validate with current code - i have to do a release before merging this
